pyro/infer/mcmc/hmc.py

```diff
@@ -188,20 +219,109 @@ def _find_reasonable_step_size(self, z):
         while direction_new == direction:
             step_size = step_size_scale * step_size
             z_new, r_new, z_grads, potential_energy = single_step_velocity_verlet(
-                z, r, self._potential_energy, step_size)
+                z, r, self._potential_energy, self._inverse_mass_matrix, step_size)
             energy_new = potential_energy + self._kinetic_energy(r_new)
             delta_energy = energy_new - energy_current
-            direction_new = 1 if target_accept_logprob < -delta_energy else -1
+            direction_new = 1 if self._direction_threshold < -delta_energy else -1
         return step_size
 
+    def _configure_adaptation(self, trace):
+        self._adapt_phase = True
+        self._adapt_window = self._adapt_initial_window
+
+        # from Stan, for small warmup_steps
+        if self._warmup_steps < 20:
+            self._adapt_start_buffer = self._warmup_steps
```

Comment:
Since we are basically not doing mass matrix adaptation when warmup steps is smaller than this, can we `return` early? This is more from the point of code readability, since attributes like `self._adapt_mass_matrix_phase_ending` will have wrongly set values (won't be used though). 

There is also a lot of state setting that is happening in this part, which might be tougher to maintain / refactor in the future. We should think about ways this can be refactored - say by constructing an adaptation schedule via `build_adaptation_schedule(warmup_steps)` which depending on the options provided returns something like `[(10, "start_buffer"), (20, "window_1"), (40, "window_2")..]` etc. Then we can peek / pop off the items from this queue and do the required operations. We can defer this to a separate refactoring PR though.

Reply:
Yup, I have thought about this for a while but did not come up with a good solution though. :(

Reply:
Np at all! It is working great, we can refactor to make things less mutable in a later PR.